File: sdk/rust/src/imessage/resources/contact_rules.rs
//! iMessage contact rules (per-identity allow/block rules + org-wide list).
//!
//! Shared iMessage pool numbers are global infrastructure, so the policy
//! owner is the agent identity being reached — rules are addressed by
//! `agent_handle`, not by a phone-number id.

use std::sync::Arc;

use serde_json::json;
use uuid::Uuid;

use crate::error::Result;
use crate::http::HttpTransport;
use crate::imessage::types::{
    ContactRuleStatus, IMessageContactRule, IMessageRuleAction, IMessageRuleMatchType,
};

const ORG_BASE: &str = "/contact-rules";

/// Build the per-identity contact-rule path, optionally addressing a rule by id.
fn rule_path(agent_handle: &str, rule_id: Option<&str>) -> String {
    let base = format!("/identities/{agent_handle}/contact-rules");
    match rule_id {
        None => base,
        Some(id) => format!("{base}/{id}"),
    }
}

/// Allow/block rules scoped to agent identities for iMessage.
pub struct IMessageContactRulesResource {
    http: Arc<HttpTransport>,
}

impl IMessageContactRulesResource {
    pub fn new(http: Arc<HttpTransport>) -> Self {
        Self { http }
    }

    /// List contact rules for one agent identity.
    ///
    /// # Arguments
    /// * `agent_handle` - Handle of the agent identity owning the rules.
    /// * `action` - Optional filter by `allow` or `block`.
    /// * `match_type` - Optional filter by `exact_number`.
    /// * `limit` - Optional max results.
    /// * `offset` - Optional pagination offset.
    pub fn list(
        &self,
        agent_handle: &str,
        action: Option<IMessageRuleAction>,
        match_type: Option<IMessageRuleMatchType>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<IMessageContactRule>> {
        // Push only present params (Python omits None keys).
        let mut params: Vec<(&str, String)> = Vec::new();
        if let Some(a) = action {
            params.push(("action", a.as_str().to_string()));
        }
        if let Some(m) = match_type {
            params.push(("match_type", m.as_str().to_string()));
        }
        if let Some(l) = limit {
            params.push(("limit", l.to_string()));
        }
        if let Some(o) = offset {
            params.push(("offset", o.to_string()));
        }
        let data = self.http.get(&rule_path(agent_handle, None), &params)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Get a single contact rule by id.
    pub fn get(&self, agent_handle: &str, rule_id: &str) -> Result<IMessageContactRule> {
        let data = self.http.get(
            &rule_path(agent_handle, Some(rule_id)),
            crate::http::NO_QUERY,
        )?;
        Ok(serde_json::from_value(data)?)
    }

    /// Create a rule. New rules are always `active`; use [`Self::update`] to
    /// pause one after creation.
    ///
    /// Returns [`crate::error::InkboxError::DuplicateContactRule`] on 409 when a
    /// non-deleted rule with the same `(match_type, match_target)` already exists.
    ///
    /// # Arguments
    /// * `agent_handle` - Handle of the agent identity owning the rule.
    /// * `action` - Whether to `allow` or `block`.
    /// * `match_target` - The value to match against (e.g. an E.164 number).
    /// * `match_type` - What to match on; defaults to `exact_number`.
    pub fn create(
        &self,
        agent_handle: &str,
        action: IMessageRuleAction,
        match_target: &str,
        match_type: IMessageRuleMatchType,
    ) -> Result<IMessageContactRule> {
        let body = json!({
            "action": action.as_str(),
            "match_type": match_type.as_str(),
            "match_target": match_target,
        });
        let data = self.http.post(
            &rule_path(agent_handle, None),
            Some(&body),
            crate::http::NO_QUERY,
        )?;
        Ok(serde_json::from_value(data)?)
    }

    /// Update `action` or `status` (admin-only).
    ///
    /// `None` arguments are omitted from the request body, mirroring the
    /// Python `_UNSET` sentinel.
    ///
    /// # Arguments
    /// * `agent_handle` - Handle of the agent identity owning the rule.
    /// * `rule_id` - Id of the rule to update.
    /// * `action` - Optional new action.
    /// * `status` - Optional new status.
    pub fn update(
        &self,
        agent_handle: &str,
        rule_id: &str,
        action: Option<IMessageRuleAction>,
        status: Option<ContactRuleStatus>,
    ) -> Result<IMessageContactRule> {
        // Build the body inserting only the fields that were supplied.
        let mut map = serde_json::Map::new();
        if let Some(a) = action {
            map.insert("action".to_string(), json!(a.as_str()));
        }
        if let Some(s) = status {
            map.insert("status".to_string(), json!(s.as_str()));
        }
        let body = serde_json::Value::Object(map);
        let data = self
            .http
            .patch(&rule_path(agent_handle, Some(rule_id)), &body)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Delete a rule (admin-only).
    pub fn delete(&self, agent_handle: &str, rule_id: &str) -> Result<()> {
        self.http.delete(&rule_path(agent_handle, Some(rule_id)))
    }

    /// Org-wide list of iMessage contact rules (admin-only).
    ///
    /// # Arguments
    /// * `agent_identity_id` - Narrow to a single agent identity by id.
    /// * `action` - Filter by `allow` or `block`.
    /// * `match_type` - Filter by `exact_number`.
    /// * `limit` - Optional max results.
    /// * `offset` - Optional pagination offset.
    pub fn list_all(
        &self,
        agent_identity_id: Option<&Uuid>,
        action: Option<IMessageRuleAction>,
        match_type: Option<IMessageRuleMatchType>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<IMessageContactRule>> {
        let mut params: Vec<(&str, String)> = Vec::new();
        if let Some(id) = agent_identity_id {
            params.push(("agent_identity_id", id.to_string()));
        }
        if let Some(a) = action {
            params.push(("action", a.as_str().to_string()));
        }
        if let Some(m) = match_type {
            params.push(("match_type", m.as_str().to_string()));
        }
        if let Some(l) = limit {
            params.push(("limit", l.to_string()));
        }
        if let Some(o) = offset {
            params.push(("offset", o.to_string()));
        }
        let data = self.http.get(ORG_BASE, &params)?;
        Ok(serde_json::from_value(data)?)
    }
}

File: sdk/typescript/src/imessage/resources/contactRules.ts
/**
 * inkbox-imessage/resources/contactRules.ts
 *
 * Per-identity iMessage contact rules (allow/block) + org-wide list.
 *
 * Shared iMessage pool numbers are global infrastructure, so the policy
 * owner is the agent identity being reached — rules are addressed by
 * `agentHandle`, not by a phone-number id.
 */

import { HttpTransport } from "../../_http.js";
import { type RuleDirection, ruleUpdateToWire } from "../../contact_rules.js";
import {
  IMessageContactRule,
  IMessageRuleAction,
  IMessageRuleMatchType,
  RawIMessageContactRule,
  parseIMessageContactRule,
} from "../types.js";

const ORG_BASE = "/contact-rules";

function rulePath(agentHandle: string, ruleId?: string): string {
  const base = `/identities/${agentHandle}/contact-rules`;
  return ruleId ? `${base}/${ruleId}` : base;
}

export interface ListIMessageContactRulesOptions {
  direction?: RuleDirection;
  action?: IMessageRuleAction;
  matchType?: IMessageRuleMatchType;
  limit?: number;
  offset?: number;
}

export interface CreateIMessageContactRuleOptions {
  direction?: RuleDirection;
  action: IMessageRuleAction;
  matchTarget: string;
  matchType?: IMessageRuleMatchType;
}

export interface UpdateIMessageContactRuleOptions {
  action?: IMessageRuleAction;
  direction?: RuleDirection;
  applyTo?: "inbound" | "outbound";
}

export interface ListAllIMessageContactRulesOptions {
  direction?: RuleDirection;
  agentIdentityId?: string;
  action?: IMessageRuleAction;
  matchType?: IMessageRuleMatchType;
  limit?: number;
  offset?: number;
}

export class IMessageContactRulesResource {
  constructor(private readonly http: HttpTransport) {}

  async list(
    agentHandle: string,
    options: ListIMessageContactRulesOptions = {},
  ): Promise<IMessageContactRule[]> {
    const params: Record<string, string | number | undefined> = {};
    if (options.direction !== undefined) params.direction = options.direction;
    if (options.action !== undefined) params.action = options.action;
    if (options.matchType !== undefined) params.match_type = options.matchType;
    if (options.limit !== undefined) params.limit = options.limit;
    if (options.offset !== undefined) params.offset = options.offset;
    const data = await this.http.get<RawIMessageContactRule[]>(
      rulePath(agentHandle),
      params,
    );
    return data.map(parseIMessageContactRule);
  }

  async get(agentHandle: string, ruleId: string): Promise<IMessageContactRule> {
    const data = await this.http.get<RawIMessageContactRule>(
      rulePath(agentHandle, ruleId),
    );
    return parseIMessageContactRule(data);
  }

  /**
   * Create a rule with an allow/block action.
   *
   * Compatible coverage may widen an existing rule and retain its ID.
   * @throws {DuplicateContactRuleError} 409 when the coverage already exists.
   */
  async create(
    agentHandle: string,
    options: CreateIMessageContactRuleOptions,
  ): Promise<IMessageContactRule> {
    const body: Record<string, unknown> = {
      action: options.action,
      match_type: options.matchType ?? IMessageRuleMatchType.EXACT_NUMBER,
      match_target: options.matchTarget,
    };
    if (options.direction !== undefined) body.direction = options.direction;
    const data = await this.http.post<RawIMessageContactRule>(
      rulePath(agentHandle),
      body,
    );
    return parseIMessageContactRule(data);
  }

  /** Update action or coverage, or atomically edit one covered side (admin-only). */
  async update(
    agentHandle: string,
    ruleId: string,
    options: UpdateIMessageContactRuleOptions,
  ): Promise<IMessageContactRule> {
    const body = ruleUpdateToWire(options);
    const data = await this.http.patch<RawIMessageContactRule>(
      rulePath(agentHandle, ruleId),
      body,
    );
    return parseIMessageContactRule(data);
  }

  /** Delete a rule (admin-only). */
  async delete(agentHandle: string, ruleId: string): Promise<void> {
    await this.http.delete(rulePath(agentHandle, ruleId));
  }

  /** Org-wide list of iMessage contact rules (admin-only). */
  async listAll(
    options: ListAllIMessageContactRulesOptions = {},
  ): Promise<IMessageContactRule[]> {
    const params: Record<string, string | number | undefined> = {};
    if (options.direction !== undefined) params.direction = options.direction;
    if (options.agentIdentityId !== undefined) params.agent_identity_id = options.agentIdentityId;
    if (options.action !== undefined) params.action = options.action;
    if (options.matchType !== undefined) params.match_type = options.matchType;
    if (options.limit !== undefined) params.limit = options.limit;
    if (options.offset !== undefined) params.offset = options.offset;
    const data = await this.http.get<RawIMessageContactRule[]>(ORG_BASE, params);
    return data.map(parseIMessageContactRule);
  }
}
